integtest/src/main/java/org/corfudb/universe/cluster/docker/CorfuServerDockerized.java

```diff
@@ -0,0 +1,324 @@
+import org.corfudb.runtime.view.Layout;
+import org.corfudb.universe.cluster.ClusterException;
+import org.corfudb.universe.node.CorfuServer;
+import org.corfudb.universe.node.NodeException;
+import org.corfudb.util.NodeLocator;
+
+import java.io.FileReader;
+import java.io.IOException;
+import java.time.Duration;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+
+import static lombok.Builder.Default;
+import static org.corfudb.runtime.CorfuRuntime.fromParameters;
+import static org.corfudb.universe.cluster.Cluster.ClusterParams;
+
+/**
+ * Implements a docker instance representing a Corfu CorfuServer.
+ */
+@Slf4j
+@Builder
+public class CorfuServerDockerized implements CorfuServer {
+    @Getter
+    private final ServerParams params;
+    private final DockerClient docker;
+    private final ClusterParams clusterParams;
+    @Default
+    private final Optional<String> containerId = Optional.empty();
+    @Default
+    private final Optional<CorfuManagementServer> runtime = Optional.empty();
+
+    /**
+     * Deploys a Corfu server / docker container
+     */
+    @Override
+    public CorfuServerDockerized deploy() throws NodeException {
+        log.info("Deploying the Corfu server. {}", params.getGenericName());
+
+        if (containerId.isPresent()) {
+            throw new IllegalStateException("Corfu server already started. Parameters: " + params);
+        }
+
+        String id = deployContainer();
+
+        return CorfuServerDockerized.builder()
+                .clusterParams(clusterParams)
+                .params(params)
+                .docker(docker)
+                .containerId(Optional.of(id))
+                .runtime(Optional.of(new CorfuManagementServer(params)))
+                .build();
+    }
+
+    @Override
+    public boolean addNode(CorfuServer server) {
+        return runtime.map(rt -> rt.add(server)).orElse(false);
+    }
+
+    @Override
+    public boolean removeNode(CorfuServer server) {
+        return runtime.map(rt -> rt.remove(server)).orElse(false);
+    }
+
+
+    @Override
+    public Optional<Layout> getLayout() {
+        return runtime.map(CorfuManagementServer::getLayout);
+    }
+
+    @Override
+    public void connectCorfuRuntime() {
+        runtime.ifPresent(CorfuManagementServer::connect);
+    }
+
+    /**
+     * This method attempts to gracefully stops the Corfu server. After timeout, it will kill the Corfu server.
+     *
+     * @param timeout a duration after which the stop will kill the server
+     * @throws NodeException this exception will be thrown if the server can not be stopped.
+     */
+    @Override
+    public void stop(Duration timeout) throws NodeException {
+        log.info("Stopping the Corfu server. {}", params.getGenericName());
+
+        try {
+            docker.stopContainer(params.getGenericName(), (int) timeout.getSeconds());
```

Comment:
Since we give a name to the container, so params.getGenericName() == containerId, so the container name can be used instead of container id. We can use the name to do everything we need (start, stop, restart, inspect and so on), we don't need to keep the container id.

Reply:
I was actually referring to the state of the container. We attempt to stop it but maybe it was already stopped.. or we could keep state that it is stopped after it is actually stopped and that could enable us to better manage action inter-dependency.

Reply:
docker provides all those features, out of the box:
`docker run -d -ti --name my ubuntu` - run an ununtu container in daemon mode.
`docker stop my`  - stop the container
`docker inspect my` - inspect the docker container
`docker start my` - start the docker container again

I mean you don't need to have the container id, to have the state of the docker container, it's enough to have a container name for that

Reply:
In the latest commit, I check that the container is not running only then start it